In [1]:
import pandas as pd
import numpy as np

/Users/anonsuksak/opt/anaconda3/lib/python3.9/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.7.3' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
/Users/anonsuksak/opt/anaconda3/lib/python3.9/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.2' currently installed).
  from pandas.core import (


In [2]:
df = pd.read_csv("2021-2022.csv")

In [8]:
df2 = df
df2

Div        Date   Time        HomeTeam        AwayTeam  FTHG  FTAG FTR  \
0    E0  13/08/2021  20:00       Brentford         Arsenal     2     0   H   
1    E0  14/08/2021  12:30      Man United           Leeds     5     1   H   
2    E0  14/08/2021  15:00         Burnley        Brighton     1     2   A   
3    E0  14/08/2021  15:00         Chelsea  Crystal Palace     3     0   H   
4    E0  14/08/2021  15:00         Everton     Southampton     3     1   H   
..   ..         ...    ...             ...             ...   ...   ...  ..   
375  E0  22/05/2022  16:00  Crystal Palace      Man United     1     0   H   
376  E0  22/05/2022  16:00       Leicester     Southampton     4     1   H   
377  E0  22/05/2022  16:00       Liverpool          Wolves     3     1   H   
378  E0  22/05/2022  16:00        Man City     Aston Villa     3     2   H   
379  E0  22/05/2022  16:00         Norwich       Tottenham     0     5   A   

     HTHG  HTAG  ... AvgC<2.5  AHCh  B365CAHH  B365CAHA  PCAHH  PCAHA  \
0       1     0  ...     1.62  0.50      1.75      2.05   1.81   2.13   
1       1     0  ...     2.25 -1.00      2.05      1.75   2.17   1.77   
2       1     0  ...     1.62  0.25      1.79      2.15   1.81   2.14   
3       2     0  ...     1.94 -1.50      2.05      1.75   2.12   1.81   
4       0     1  ...     1.67 -0.50      2.05      1.88   2.05   1.88   
..    ...   ...  ...      ...   ...       ...       ...    ...    ...   
375     1     0  ...     2.04  0.25      1.68      2.15   1.74   2.23   
376     0     0  ...     2.63 -0.75      1.83      2.07   1.88   2.03   
377     1     1  ...     3.28 -2.50      2.02      1.77   2.06   1.83   
378     0     1  ...     3.36 -2.25      2.06      1.84   2.05   1.86   
379     0     2  ...     2.40  1.50      1.95      1.95   1.96   1.94   

     MaxCAHH  MaxCAHA  AvgCAHH  AvgCAHA  
0       2.05     2.17     1.80     2.09  
1       2.19     1.93     2.10     1.79  
2       1.82     2.19     1.79     2.12  
3       2.16     1.93     2.06     1.82  
4       2.08     1.90     2.03     1.86  
..       ...      ...      ...      ...  
375     1.88     2.25     1.74     2.16  
376     1.94     2.26     1.87     2.01  
377     2.19     1.99     2.07     1.80  
378     2.09     2.03     2.01     1.87  
379     2.16     1.98     1.99     1.89  

[380 rows x 106 columns]

## เนื่องจากต้องการใช้ Year จึงเขียนโค้ดเพื่อแยก Year ออกจาก column "Date"

In [9]:
# Assuming df is your DataFrame and the date column is named 'Date'
df2['Date'] = pd.to_datetime(df2['Date'])
df2['Year'] = df2['Date'].dt.year

/var/folders/kn/8gjhms0n1pngh9vl05r8zqg80000gn/T/ipykernel_53597/2769034650.py:2: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df2['Date'] = pd.to_datetime(df2['Date'])


In [10]:
df2

Div       Date   Time        HomeTeam        AwayTeam  FTHG  FTAG FTR  \
0    E0 2021-08-13  20:00       Brentford         Arsenal     2     0   H   
1    E0 2021-08-14  12:30      Man United           Leeds     5     1   H   
2    E0 2021-08-14  15:00         Burnley        Brighton     1     2   A   
3    E0 2021-08-14  15:00         Chelsea  Crystal Palace     3     0   H   
4    E0 2021-08-14  15:00         Everton     Southampton     3     1   H   
..   ..        ...    ...             ...             ...   ...   ...  ..   
375  E0 2022-05-22  16:00  Crystal Palace      Man United     1     0   H   
376  E0 2022-05-22  16:00       Leicester     Southampton     4     1   H   
377  E0 2022-05-22  16:00       Liverpool          Wolves     3     1   H   
378  E0 2022-05-22  16:00        Man City     Aston Villa     3     2   H   
379  E0 2022-05-22  16:00         Norwich       Tottenham     0     5   A   

     HTHG  HTAG  ...  AHCh B365CAHH  B365CAHA  PCAHH  PCAHA  MaxCAHH  MaxCAHA  \
0       1     0  ...  0.50     1.75      2.05   1.81   2.13     2.05     2.17   
1       1     0  ... -1.00     2.05      1.75   2.17   1.77     2.19     1.93   
2       1     0  ...  0.25     1.79      2.15   1.81   2.14     1.82     2.19   
3       2     0  ... -1.50     2.05      1.75   2.12   1.81     2.16     1.93   
4       0     1  ... -0.50     2.05      1.88   2.05   1.88     2.08     1.90   
..    ...   ...  ...   ...      ...       ...    ...    ...      ...      ...   
375     1     0  ...  0.25     1.68      2.15   1.74   2.23     1.88     2.25   
376     0     0  ... -0.75     1.83      2.07   1.88   2.03     1.94     2.26   
377     1     1  ... -2.50     2.02      1.77   2.06   1.83     2.19     1.99   
378     0     1  ... -2.25     2.06      1.84   2.05   1.86     2.09     2.03   
379     0     2  ...  1.50     1.95      1.95   1.96   1.94     2.16     1.98   

     AvgCAHH  AvgCAHA  Year  
0       1.80     2.09  2021  
1       2.10     1.79  2021  
2       1.79     2.12  2021  
3       2.06     1.82  2021  
4       2.03     1.86  2021  
..       ...      ...   ...  
375     1.74     2.16  2022  
376     1.87     2.01  2022  
377     2.07     1.80  2022  
378     2.01     1.87  2022  
379     1.99     1.89  2022  

[380 rows x 107 columns]

## Set question and answer (implement)
## 1. Simple stats 3 questions 

## 1.1 How many teams have been the home team, and how many times have they been the home team?

In [11]:
mhomet = df2[["HomeTeam", "Year"]].nunique()
print(f"{mhomet['HomeTeam']} teams have been HomeTeam within 2021-2022.")

mhomet2 = df2["HomeTeam"].value_counts()
mhomet2

20 teams have been HomeTeam within 2021-2022.


HomeTeam
Brentford         19
Man United        19
Arsenal           19
Wolves            19
Southampton       19
Brighton          19
Man City          19
Leeds             19
Crystal Palace    19
Aston Villa       19
Liverpool         19
Tottenham         19
Newcastle         19
Norwich           19
Watford           19
Leicester         19
Everton           19
Chelsea           19
Burnley           19
West Ham          19
Name: count, dtype: int64

- ทีมเจ้าบ้านมีทั้งหมด 20 ทีม ที่มีการแข่งขันในปี 2021-2022 และแต่ละทีมที่แข่งขัน มีโอกาสเป็นเจ้าบ้าน (HomeTeam) เท่ากันที่ 19 ครั้ง

## 1.2 How many times have they been the home team in 2021 and 2022, and which team was the most home team in each year?

In [12]:
df2.rename(columns={'Date': 'Times'}, inplace=True)

## Step 1 How many times have they been the home team in 2021 and 2022?

In [13]:
df2.groupby(["HomeTeam", "Year"])["Times"].count()

HomeTeam        Year
Arsenal         2021     9
                2022    10
Aston Villa     2021     9
                2022    10
Brentford       2021     9
                2022    10
Brighton        2021     9
                2022    10
Burnley         2021     7
                2022    12
Chelsea         2021    10
                2022     9
Crystal Palace  2021    10
                2022     9
Everton         2021     8
                2022    11
Leeds           2021     9
                2022    10
Leicester       2021     9
                2022    10
Liverpool       2021     9
                2022    10
Man City        2021    10
                2022     9
Man United      2021     9
                2022    10
Newcastle       2021    10
                2022     9
Norwich         2021    10
                2022     9
Southampton     2021     9
                2022    10
Tottenham       2021    10
                2022     9
Watford         2021     9
                2022    10
West Ha

## Step 2 Which team was the most home team in each year?

In [38]:
result = df2.groupby(["Year", "HomeTeam"]).count().reset_index()[["Year", "HomeTeam", "Times"]]

# Find the index of the maximum count for each year
idx_max = result.groupby('Year')['Times'].idxmax()

# Filter the original DataFrame based on the indices of the maximum counts
most_home_team_each_year = result.loc[idx_max]

most_home_team_each_year

Year HomeTeam  Times
5   2021  Chelsea     10
24  2022  Burnley     12

- โดยส่วนใหญ่แล้ว แต่ละทีมที่แข่งขัน จะเป็นเจ้าบ้าน ในปี 2021 ทีมละ 9 ครั้ง และปี 2022 ทีมละ 10 ครั้ง ยกเว้นทีม Burnley ที่ปี 2021 เป็นเจ้าบ้าน 7 ครั้ง ปี 2022 เป็นเจ้าบ้าน 12 ครั้ง ทีม Crystal Palce ปี 2021 เป็นเจ้าบ้าน 10 ครั้ง ปี 2022 เป็นเจ้าบ้าน 9 ครั้ง และทีม Everton เป็นเจ้าบ้าน 8 ครั้ง (2021) และปี 2022 จำนวน 11 ครั้ง เป็นต้น
- ซึ่งโดยภาพรวมแล้ว ในปี 2021 ทีม Chelsea เป็นเจ้าบ้านมากที่สุด จำนวน 10 ครั้ง และปี 2022 ทีม Burnley เป็นเจ้าบ้านมากถึง 12 ครั้ง

## 1.3 How many goals did the home team score and what about away team?  (Descending)

In [22]:
FTHTG = df2.groupby("HomeTeam")["FTHG"].sum()
FTHTG_sorted = FTHTG.sort_values(ascending=False)
FTHTG_sorted

HomeTeam
Man City          58
Liverpool         49
Tottenham         38
Chelsea           37
Arsenal           35
Leicester         34
West Ham          33
Man United        32
Aston Villa       29
Crystal Palace    27
Everton           27
Newcastle         26
Southampton       23
Brentford         22
Wolves            20
Leeds             19
Brighton          19
Burnley           18
Watford           17
Norwich           12
Name: FTHG, dtype: int64

In [23]:
FTATG = df2.groupby("AwayTeam")["FTAG"].sum()
FTATG_sorted = FTATG.sort_values(ascending=False)
FTATG_sorted

AwayTeam
Liverpool         45
Man City          41
Chelsea           39
Tottenham         31
Leicester         28
West Ham          27
Arsenal           26
Brentford         26
Man United        25
Leeds             23
Crystal Palace    23
Aston Villa       23
Brighton          23
Southampton       20
Newcastle         18
Wolves            18
Watford           17
Everton           16
Burnley           16
Norwich           11
Name: FTAG, dtype: int64

- กรณีเจ้าบ้าน พบว่า ทีม man city สามารถทำประตูได้มากที่สุด จำนวน 58 ประตู รองลงมาคือ Liverpool (49 ประตู) และ Tottenham (38 ประตู)
- กรณีเป็นทีมเยื่อน พบว่า ทีม Liverpool สามารถทำประตูได้มากที่สุด จำนวน 45 ประตู รองลงมาคือ man city (41 ประตู) และ Chelsea (39 ประตู)
- ส่วนทีมที่ทำประตูได้น้อยที่สุด ทั้งกรณีเป็นเจ้าบ้าน (12 ประตู) และทีมเยื่อน (11 ประตู) คือ Norwich

## 2. Advanced 2 questions

## 2.1 Top 5 HomeTeams with the lowest lose rates

In [28]:
top5L = df2.groupby(["HomeTeam","FTR"])["FTR"].count().unstack(fill_value=0)
top5L

FTR              A  D   H
HomeTeam                 
Arsenal          4  2  13
Aston Villa      8  5   6
Brentford        9  3   7
Brighton         7  7   5
Burnley          8  6   5
Chelsea          3  7   9
Crystal Palace   4  8   7
Everton          8  2   9
Leeds            9  6   4
Leicester        5  4  10
Liverpool        0  4  15
Man City         2  2  15
Man United       4  5  10
Newcastle        5  6   8
Norwich         13  3   3
Southampton      6  7   6
Tottenham        5  1  13
Watford         15  2   2
West Ham         5  5   9
Wolves           9  3   7

In [29]:
# Calculate the percentage for each 'FTR' category
top5L_ = top5L.div(top5L.sum(axis=1), axis=0)
top5L_

FTR                    A         D         H
HomeTeam                                    
Arsenal         0.210526  0.105263  0.684211
Aston Villa     0.421053  0.263158  0.315789
Brentford       0.473684  0.157895  0.368421
Brighton        0.368421  0.368421  0.263158
Burnley         0.421053  0.315789  0.263158
Chelsea         0.157895  0.368421  0.473684
Crystal Palace  0.210526  0.421053  0.368421
Everton         0.421053  0.105263  0.473684
Leeds           0.473684  0.315789  0.210526
Leicester       0.263158  0.210526  0.526316
Liverpool       0.000000  0.210526  0.789474
Man City        0.105263  0.105263  0.789474
Man United      0.210526  0.263158  0.526316
Newcastle       0.263158  0.315789  0.421053
Norwich         0.684211  0.157895  0.157895
Southampton     0.315789  0.368421  0.315789
Tottenham       0.263158  0.052632  0.684211
Watford         0.789474  0.105263  0.105263
West Ham        0.263158  0.263158  0.473684
Wolves          0.473684  0.157895  0.368421

In [30]:
# Display the result
top5L_ = top5L_.sort_values(by="A", ascending=True).head()
top5L_["A"]

HomeTeam
Liverpool     0.000000
Man City      0.105263
Chelsea       0.157895
Arsenal       0.210526
Man United    0.210526
Name: A, dtype: float64

- จากข้อมูล 2021-2022 พบว่า หากทีม Liverpool เป็นเจ้าบ้าน อัตราการแพ้จะเท่ากับ 0 หมายความว่า แข่งทั้งหมด 19 นัด (เจ้าบ้าน) จะชนะทั้ง 19 นัด
- รองลงมาคือ ทีม Man City, Chelsea, Arsenal และ Man United ที่มีอัตราการแพ้ต่ำที่สุด 5 ลำดับแรก (กรณีเป็นเจ้าบ้าน)

## 2.2 Did obtaining yellow and red cards affect the home team's win rate?

## Step 1 กำหนดให้ H = 1, และ D, A = 0 เพื่อจะวิเคราะห์โอกาสการชนะ และทำ groupby เพื่อดูข้อมูล

In [31]:
df2['HomeWin'] = df2['FTR'].apply(lambda result: 1 if result == 'H' else 0)

In [32]:
df2

Div      Times   Time        HomeTeam        AwayTeam  FTHG  FTAG FTR  \
0    E0 2021-08-13  20:00       Brentford         Arsenal     2     0   H   
1    E0 2021-08-14  12:30      Man United           Leeds     5     1   H   
2    E0 2021-08-14  15:00         Burnley        Brighton     1     2   A   
3    E0 2021-08-14  15:00         Chelsea  Crystal Palace     3     0   H   
4    E0 2021-08-14  15:00         Everton     Southampton     3     1   H   
..   ..        ...    ...             ...             ...   ...   ...  ..   
375  E0 2022-05-22  16:00  Crystal Palace      Man United     1     0   H   
376  E0 2022-05-22  16:00       Leicester     Southampton     4     1   H   
377  E0 2022-05-22  16:00       Liverpool          Wolves     3     1   H   
378  E0 2022-05-22  16:00        Man City     Aston Villa     3     2   H   
379  E0 2022-05-22  16:00         Norwich       Tottenham     0     5   A   

     HTHG  HTAG  ... B365CAHH B365CAHA  PCAHH  PCAHA  MaxCAHH  MaxCAHA  \
0       1     0  ...     1.75     2.05   1.81   2.13     2.05     2.17   
1       1     0  ...     2.05     1.75   2.17   1.77     2.19     1.93   
2       1     0  ...     1.79     2.15   1.81   2.14     1.82     2.19   
3       2     0  ...     2.05     1.75   2.12   1.81     2.16     1.93   
4       0     1  ...     2.05     1.88   2.05   1.88     2.08     1.90   
..    ...   ...  ...      ...      ...    ...    ...      ...      ...   
375     1     0  ...     1.68     2.15   1.74   2.23     1.88     2.25   
376     0     0  ...     1.83     2.07   1.88   2.03     1.94     2.26   
377     1     1  ...     2.02     1.77   2.06   1.83     2.19     1.99   
378     0     1  ...     2.06     1.84   2.05   1.86     2.09     2.03   
379     0     2  ...     1.95     1.95   1.96   1.94     2.16     1.98   

     AvgCAHH  AvgCAHA  Year  HomeWin  
0       1.80     2.09  2021        1  
1       2.10     1.79  2021        1  
2       1.79     2.12  2021        0  
3       2.06     1.82  2021        1  
4       2.03     1.86  2021        1  
..       ...      ...   ...      ...  
375     1.74     2.16  2022        1  
376     1.87     2.01  2022        1  
377     2.07     1.80  2022        1  
378     2.01     1.87  2022        1  
379     1.99     1.89  2022        0  

[380 rows x 108 columns]

In [34]:
win_ = df2.groupby("HomeTeam")[["HY","HR","HomeWin"]].sum()

win_.sort_values(by=["HY", "HR"], ascending=False) 

#HY = Hometeam yellow card
#HR = Hometeam red card

HY  HR  HomeWin
HomeTeam                       
Leeds           53   2        4
Everton         41   5        9
Newcastle       34   1        8
Burnley         34   0        5
Crystal Palace  34   0        7
Tottenham       34   0       13
Watford         33   1        2
Brighton        32   1        5
Chelsea         32   0        9
Aston Villa     31   2        6
Arsenal         31   1       13
Man United      31   1       10
Southampton     31   1        6
Wolves          31   1        7
Brentford       29   2        7
Norwich         29   0        3
West Ham        27   0        9
Liverpool       23   0       15
Leicester       22   0       10
Man City        16   1       15

## Step 2 วิเคราะห์ Regression โดยต้องการดูว่า การได้รับใบเหลือง หรือใบแดง หรือทั้งสองอย่าง จะกระทบต่ออัตราการชนะของทีมเจ้าบ้านหรือไม่
## Model: lnHometeamWin = beta0 + beta1HY + beta2HR

In [35]:
import numpy as np
import statsmodels.api as sm
import pandas as pd

In [36]:
# Create the dependent variable (y) and independent variables (X)
y = np.log1p(df2['HomeWin'])
X = df2[['HY', 'HR']]

# Add a constant term to the independent variables
X = sm.add_constant(X)

# Fit the regression model
model = sm.OLS(y, X).fit()

# Print the regression results
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                HomeWin   R-squared:                       0.058
Model:                            OLS   Adj. R-squared:                  0.053
Method:                 Least Squares   F-statistic:                     11.63
Date:                Sun, 04 Feb 2024   Prob (F-statistic):           1.25e-05
Time:                        21:20:39   Log-Likelihood:                -121.27
No. Observations:                 380   AIC:                             248.5
Df Residuals:                     377   BIC:                             260.4
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.3882      0.028     13.647      0.0

## อีกโค้ด ที่ใช้ Sklearn

In [37]:
import numpy as np
from sklearn.linear_model import LinearRegression
import pandas as pd

# Assuming df2 is your DataFrame
X = df2[['HY', 'HR']]
Y = np.log(df2['HomeWin'] + 1)  # Adding 1 to handle cases where 'HomeWin' is zero

# Create a linear regression model
model = LinearRegression()

# Fit the model
model.fit(X, Y)

# Print the coefficients
print('Intercept:', model.intercept_)
print('Coefficients:', model.coef_)

Intercept: 0.3882067462413569
Coefficients: [-0.04791944 -0.23379818]


## Step 3 ผลการวิเคราะห์ 
## logHometeamwin = 0.3882 - 0.0479HY - 0.2338HR
- ผลการศึกษา พบว่า ค่าสัมประสิทธิ์ของทั้ง HR (Hometeam red card), HY (Hometeam yellow card) มีค่าติดลบ ดังนั้นหาก obtaining yellow and red cards จะทำให้โอกาสในการชนะลดลง
- เช่น หาก Hometeam ได้รับ Red card 0 ใบ Yellow card 0 ใบ โอกาสชนะจะเท่ากับ HomeWin = 0.3882-0.0479(0) - 0.2338(0) = 0.3882%
- แต่หากได้รับ Red card 1 ใบ Yellow card 1 ใบ โอกาสชนะจะลดลงเป็น HomeWin = 0.3882-0.0479(1) - 0.2338(1) = 0.1065%
- ซึ่งยิ่งได้ Red card และ Yellow card มากขึ้น โอกาสการชนะจะยิ่งลดลง โดย Red card ทำให้โอกาสชนะลดลง มากกว่า Yellow card